lib/tasks/TaskScheduler.js

```diff
@@ -81,11 +101,53 @@ class TaskScheduler {
             }
             if (queueKey !== undefined) {
                 const queue = this._getTaskQueue(ctx, queueKey);
+                this._inQueue++;
                 return queue.push(ctx, onTaskEnd);
             }
         }
+        this._running++;
         return process.nextTick(() => this._processingFunc(ctx, onTaskEnd));
     }
+
+    /**
+     * @returns {Number} The number of tasks waiting to be processed
+     */
+    length() {
+        return this._inQueue;
+    }
+
+    /**
+     * @returns {Number} The number of tasks currently being processed
+     */
+    running() {
+        return this._running;
+    }
+
+    /**
+     * @returns {Boolean} True if there are no tasks running or waiting
+     */
+    idle() {
+        return this._running === 0 && this._inQueue === 0;
+    }
+
+    /**
+     * Set a function to be called after all tasks have been processed
+     * @param {DrainFunction} func drain function
+     * @returns {undefined}
+     */
+    drain(func) {
```

Comment:
should this be `setDrain`? drain alone makes me think this function is supposed to actually drain